### PR TITLE
Feature/upsert approved purposes

### PR DESCRIPTION
**Pull request by Gubbsy**
## Proposed changes

Upsert approved purposes and outcomes based on answers provided

### CODE REVIEW

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb [hunk 1]
```diff
@@ -284,6 +284,192 @@ def upsert_project_update(funding_application)
...
+            External_Id__c: approved_purpose.id,
+            Approved_Purpose__c: approved_purpose.description,
+            Form__c: form_id,
+            Update__c: approved_purpose.progress,
+          )
+        end
+
+        Rails.logger.info("Successfuly upserted progress update  " \
+          "purpose update with ID: #{progress_update.id}")
+
```
Comment (ptrelease) [inline lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:319]: add the word approved in log line,

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb [hunk 2]
```diff
@@ -284,6 +284,192 @@ def upsert_project_update(funding_application)
...
+        if retry_number < MAX_RETRIES
+
+          retry_number += 1
+
+          max_sleep_seconds = Float(2 ** retry_number)
+
+          Rails.logger.error(
+            "Error upserting progress update appproved purpose " \
+              "with ID: #{progress_update.id}"
+          )
```
Comment (ptrelease) [inline lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:332]: logging the progress update id, not approved purpose id

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb [hunk 3]
```diff
@@ -284,6 +284,192 @@ def upsert_project_update(funding_application)
...
+    # @param [int] form_id Form ID to upsert against
+    #                    
+    def upsert_approved_purposes(progress_update, form_id)
+
+      retry_number = 0
+
+      begin
+
+        Rails.logger.info("Upserting progress update approved purpose " \
+          "with ID: #{progress_update.id}")
```
Comment (ptrelease) [inline lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:302]: did you want to upload the approved purpose id here?

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb [hunk 4]
```diff
@@ -284,6 +284,192 @@ def upsert_project_update(funding_application)
...
+    #    
+    def upsert_outcomes(funding_application)
+      retry_number = 0
+
+      begin
+
+        progress_update = funding_application.arrears_journey_tracker.progress_update
+
+        Rails.logger.info("Upserting progress update outcomes " \
+          "with ID: #{progress_update.id}")
```
Comment (ptrelease) [inline lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:361]: quick check that this is the id you want

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb [hunk 5]
```diff
@@ -284,6 +284,192 @@ def upsert_project_update(funding_application)
...
+
+        # Upsert new project completion date if provided
+        unless progress_update.progress_update_demographic.empty?
+          @client.upsert!(
+            'Forms__c',
+            'Frontend_External_Id__c',
+            Case__c: funding_application.salesforce_case_id,
+            Frontend_External_Id__c: progress_update.id,
```
Comment (ptrelease) [inline lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:369]: Is this the right id?

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb [hunk 6]
```diff
@@ -284,6 +284,192 @@ def upsert_project_update(funding_application)
...
+              @client.upsert!(
+                'Forms__c',
+                'Frontend_External_Id__c',
+                Case__c: funding_application.salesforce_case_id,
+                Frontend_External_Id__c: progress_update.id,
+                The_organisation_will_be_more_resilient__c: update
+              )
+            end
+          end
+        end
```
Comment (ptrelease) [inline lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:444]: Love the use of the k, v pairs here.

Does mean you could abstract this:

```
              @client.upsert!(
                'Forms__c',
                'Frontend_External_Id__c',
                Case__c: funding_application.salesforce_case_id,
                Frontend_External_Id__c: progress_update.id,
                A_better_place_to_live_work_or_visit__c: update
              )
```

to its own function if you wanted.  Would save about 50 lines of code.

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb [hunk 7]
```diff
@@ -284,6 +284,192 @@ def upsert_project_update(funding_application)
...
+
+        Rails.logger.info("Successfuly upserted progress update  " \
+          "outcomes with ID: #{progress_update.id}")
+
```
Comment (ptrelease) [inline lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:448]: Be good to have a log for each one you do (you could use the salesforce field you are updating to make unique).  Then we can see from the logs what went to Salesforce.  Consider this if you like the function idea above - as that will make this easy.

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb (hunk 6 above)
Comment (Gubbsy) [reply lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:444]: Can't have this as an abstracted function as the unique key/symbol for the SF field to upload against can't be passed into a method to call on client.upsert(). Same issue we had with having a very verbose upload progress update method

File: lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb (hunk 2 above)
Comment (Gubbsy) [reply lib/apis/salesforce/progress_update/progress_update_salesforce_api.rb:333]: I have no way of accessing the approved purpose ID from this rescue?

**Review by ptrelease — APPROVED** [review]

**Pull request merged**